src/python_testing/TC_DISHM_3_1.py

```diff
@@ -0,0 +1,170 @@
+        asserts.assert_true(self.check_pics("DISHM.S.A0001"), "DISHM.S.A0001 must be supported")
+        asserts.assert_true(self.check_pics("DISHM.S.C00.Rsp"), "DISHM.S.C00.Rsp must be supported")
+        asserts.assert_true(self.check_pics("DISHM.S.C01.Tx"), "DISHM.S.C01.Tx must be supported")
+
+        if not self.check_pics("DISHM.S.A0003"):
+            logging.info("Test skipped because PICS DISHM.S.A0003 (OnMode) is not set")
+            return
+
+        if not self.check_pics("DISHM.S.F00"):
+            logging.info("Test skipped because PICS DISHM.S.F00 (DepOnOff) is not set")
+            return
+
+        ret = await self.check_preconditions(self.endpoint)
+        asserts.assert_true(ret, "invalid preconditions - OnMode is non-null or not a mode in the Supported Modes attribute")
+
+        attributes = Clusters.DishwasherMode.Attributes
+
+        from enum import Enum
+
+        class CommonCodes(Enum):
+            SUCCESS = 0x00
+            UNSUPPORTED_MODE = 0x01
+            GENERIC_FAILURE = 0x02
+
+        self.print_step(1, "Commissioning, already done")
+
+        self.print_step(2, "Read OnMode attribute")
+
+        on_mode_dut = await self.read_mod_attribute_expect_success(endpoint=self.endpoint, attribute=attributes.OnMode)
+
+        logging.info("OnMode: %s" % (on_mode_dut))
+
+        asserts.assert_false(on_mode_dut == NullValue, "On mode value should be an integer value")
+
+        self.print_step(3, "Read CurrentMode attribute")
+
+        old_current_mode_dut = await self.read_mod_attribute_expect_success(endpoint=self.endpoint, attribute=attributes.CurrentMode)
+
+        logging.info("CurrentMode: %s" % (old_current_mode_dut))
+
+        if old_current_mode_dut == on_mode_dut:
+
+            self.print_step(4, "Read SupportedModes attribute")
+            supported_modes = await self.read_mod_attribute_expect_success(endpoint=self.endpoint, attribute=attributes.SupportedModes)
+
+            logging.info("SupportedModes: %s" % (supported_modes))
+
+            asserts.assert_greater_equal(len(supported_modes), 2, "SupportedModes must have at least two entries!")
+
+            new_mode_th = None
+
+            for m in supported_modes:
+                if m.mode != on_mode_dut:
+                    new_mode_th = m.mode
+                    break
+
+            self.print_step(5, "Send ChangeToMode command with NewMode set to %d" % (new_mode_th))
+
+            ret = await self.send_change_to_mode_cmd(newMode=new_mode_th)
+            asserts.assert_true(ret.status == CommonCodes.SUCCESS.value, "Changing the mode should succeed")
+
+        self.print_step(6, "Send Off command")
+
+        ret = await self.write_onoff_off_mode()
+        asserts.assert_true(ret.status == CommonCodes.SUCCESS.value, "Changing the mode to off should succeed")
+
+        self.print_step(7, "Send On command")
+
+        ret = await self.write_onoff_on_mode()
```

Comment:
There is no write_onoff_on_mode function.

Stopping here, at the end of TC_DISHM_3_1.py.  Please fix the nonexistent functions, and make sure the tests can actually run, so I am not trying to manually catch in review all the issues just running the test would catch.